sasdata/data_util/nxsunit.py

```diff
@@ -43,173 +43,405 @@
+SHORT_PREFIX = dict(P=1e15, T=1e12, G=1e9, M=1e6, k=1e3, d=1e-1, c=1e-2, m=1e-3, u=1e-6, n=1e-9, p=1e-12, f=1e-15)
 
-__all__ = ['Converter']
 
 # Limited form of units for returning objects of a specific type.
 # Maybe want to do full units handling with e.g., pyre's
 # unit class. For now lets keep it simple.  Note that
-def _build_metric_units(unit,abbr):
+def _build_metric_units(unit: str, abbr: str) -> Dict[str, float]:
     """
     Construct standard SI names for the given unit.
     Builds e.g.,
-        s, ns
-        second, nanosecond, nano*second
-        seconds, nanoseconds
+        s, ns, n*s, n_s
+        second, nanosecond, nano*second, nano_second
+        seconds, nanoseconds, nano*seconds, nano_seconds
     Includes prefixes for femto through peta.
 
     Ack! Allows, e.g., Coulomb and coulomb even though Coulomb is not
     a unit because some NeXus files store it that way!
 
     Returns a dictionary of names and scales.
     """
-    prefix = dict(peta=1e15,tera=1e12,giga=1e9,mega=1e6,kilo=1e3,
-                  deci=1e-1,centi=1e-2,milli=1e-3,mili=1e-3,micro=1e-6,
-                  nano=1e-9,pico=1e-12,femto=1e-15)
-    short_prefix = dict(P=1e15,T=1e12,G=1e9,M=1e6,k=1e3,
-                        d=1e-1,c=1e-2,m=1e-3,u=1e-6,
-                        n=1e-9,p=1e-12,f=1e-15)
-    map = {abbr:1}
-    map.update([(P+abbr,scale) for (P,scale) in short_prefix.items()])
-    for name in [unit,unit.capitalize()]:
-        map.update({name:1,name+'s':1})
-        map.update([(P+name,scale) for (P,scale) in prefix.items()])
-        map.update([(P+'*'+name,scale) for (P,scale) in prefix.items()])
-        map.update([(P+name+'s',scale) for (P,scale) in prefix.items()])
-    return map
-
-def _build_plural_units(**kw):
+    units = {}
+    for name in [unit, unit.capitalize(), unit.lower(), abbr]:
+        for items in [PREFIX, SHORT_PREFIX]:
+            names = {}
+            names.update({name: 1})
+            names.update([(P + name, scale) for (P, scale) in items.items()])
+            names.update([(P + '*' + name, scale) for (P, scale) in items.items()])
+            names.update([(P + '_' + name, scale) for (P, scale) in items.items()])
+            # Exclude pluralized abbrevs., e.g. create m(illi)(?)(*|_)(?)meters, but not milli(*|_)(?)ms or m(*|_)(?)ms
+            if name != abbr:
+                names.update(_build_plural_units(**names))
+            units.update(names)
+    return units
+
+
+def _build_plural_units(**kw: Dict[str, ConversionType]) -> Dict[str, ConversionType]:
     """
     Construct names for the given units.  Builds singular and plural form.
     """
-    map = {}
-    map.update([(name,scale) for name,scale in kw.items()])
-    map.update([(name+'s',scale) for name,scale in kw.items()])
-    return map
+    units = {}
+    units.update([(name, scale) for name, scale in kw.items()])
+    units.update([(name + 's', scale) for name, scale in kw.items()])
```

Comment:
I agree, but this class works mostly in SI units. Here is a list of edge cases we might come up against where this fails, based on my understanding of [the NIST guide to SI](https://www.nist.gov/pml/special-publication-811/nist-guide-si-chapter-9-rules-and-style-conventions-spelling-unit-names):

- Anything Imperial that isn't a temperature
  - Potential solution: add imperial lengths and weights
  - What I am willing to do for this PR: I will continue to omit this. The scattering community doesn't work in inch(es)/feet/pounds/etc.
- `microohm` and `kiloohm` will be generated by the class, but `microhm` and `kilohm` are the correct forms.
  - Potential solution: explicitly add these as resistance units
  - What I am willing to do for this PR: I could implement this pretty quickly if there is a call for it.
- units in long-form, e.g. `meters per second squared` -or- `meters per square second`
  - Potential solution 1: convert `<square|cubic> <unit>` or `<unit> <squared|cubed>` to `unit^<2|3>`
  - Potential solution 2: convert `per` to `/`
  - What I am willing to do for this PR: I ignored this because long-form is typically reserved for textual content, not data units, but implementation would be fairly quick. I'd also need to add some reasonable unit tests.